File: partiql-ast/src/main/java/org/partiql/ast/v1/DDL.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;

/**
 * TODO docs, equals, hashcode
 */
public abstract class DDL extends Statement {
    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        if (this instanceof CreateTable) {
            return visitor.visitCreateTable((CreateTable) this, ctx);
        } else if (this instanceof CreateIndex) {
            return visitor.visitCreateIndex((CreateIndex) this, ctx);
        } else if (this instanceof DropTable) {
            return visitor.visitDropTable((DropTable) this, ctx);
        } else if (this instanceof DropIndex) {
            return visitor.visitDropIndex((DropIndex) this, ctx);
        } else {
            throw new IllegalStateException("Unexpected value: " + this);
        }
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/DDL.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public abstract class DDL : Statement() {
    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = when (this) {
        is CreateTable -> visitor.visitCreateTable(this, ctx)
        is CreateIndex -> visitor.visitCreateIndex(this, ctx)
        is DropTable -> visitor.visitDropTable(this, ctx)
        is DropIndex -> visitor.visitDropIndex(this, ctx)
        else -> throw NotImplementedError()
    }
}
